paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MonitorSource.java

```diff
@@ -71,74 +71,88 @@
-                                new ListStateDescriptor<>(
-                                        "next-snapshot", LongSerializer.INSTANCE));
-
-        @SuppressWarnings("unchecked")
-        final Class<Tuple2<Long, Long>> typedTuple =
-                (Class<Tuple2<Long, Long>>) (Class<?>) Tuple2.class;
-        this.nextSnapshotState =
-                context.getOperatorStateStore()
-                        .getListState(
-                                new ListStateDescriptor<>(
-                                        "next-snapshot-per-checkpoint",
-                                        new TupleSerializer<>(
-                                                typedTuple,
-                                                new TypeSerializer[] {
-                                                    LongSerializer.INSTANCE, LongSerializer.INSTANCE
-                                                })));
-
-        this.nextSnapshotPerCheckpoint = new TreeMap<>();
-
-        if (context.isRestored()) {
-            LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+    public Boundedness getBoundedness() {
+        return Boundedness.CONTINUOUS_UNBOUNDED;
+    }
+
+    @Override
+    public SourceReader<Split, SimpleSourceSplit> createReader(
+            SourceReaderContext sourceReaderContext) throws Exception {
+        return new Reader();
+    }
+
+    private class Reader extends AbstractNonCoordinatedSourceReader<Split> {
+        private static final String CHECKPOINT_STATE = "CS";
+        private static final String NEXT_SNAPSHOT_STATE = "NSS";
+
+        private final StreamTableScan scan = readBuilder.newStreamScan();
+        private final TreeMap<Long, Long> nextSnapshotPerCheckpoint = new TreeMap<>();
 
-            List<Long> retrievedStates = new ArrayList<>();
-            for (Long entry : this.checkpointState.get()) {
-                retrievedStates.add(entry);
+        @Override
+        public void notifyCheckpointComplete(long checkpointId) {
+            NavigableMap<Long, Long> nextSnapshots =
+                    nextSnapshotPerCheckpoint.headMap(checkpointId, true);
+            OptionalLong max = nextSnapshots.values().stream().mapToLong(Long::longValue).max();
+            max.ifPresent(scan::notifyCheckpointComplete);
+            nextSnapshots.clear();
+        }
+
+        @Override
+        public List<SimpleSourceSplit> snapshotState(long checkpointId) {
+            List<SimpleSourceSplit> results = new ArrayList<>();
+
+            Long nextSnapshot = this.scan.checkpoint();
+            if (nextSnapshot != null) {
+                results.add(new SimpleSourceSplit(CHECKPOINT_STATE + nextSnapshot));
+                this.nextSnapshotPerCheckpoint.put(checkpointId, nextSnapshot);
+            }
+
+            this.nextSnapshotPerCheckpoint.forEach(
+                    (k, v) ->
+                            results.add(new SimpleSourceSplit(NEXT_SNAPSHOT_STATE + k + ":" + v)));
```

Comment:
So you're storing state information in splits? Create a utility class for this, for example:

```java
stateBuilder.storeSingle(CHECKPOINT_STATE, nextSnapshot);
stateBuilder.storeList(NEXT_SNAPSHOT_STATE, this.nextSnapshotPerCheckpoint.stream().map((k, v) -> k + ":" + v).collect(Collectors.asList()));
return stateBuilder.build();
```

And extract state with

```java
stateBuilder.getSingle(CHECKPOINT_STATE);
stateBuilder.getList(NEXT_SNAPSHOT_STATE);
```